package com.ecomm.product.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.ecomm.product.dto.BrandDto;
import com.ecomm.product.request.AddBrandRequest;
import com.ecomm.product.request.UpdateBrandRequest;
import com.ecomm.product.response.ApiResponse;
import com.ecomm.product.service.BrandService;

@RestController
@RequestMapping("/brand")
public class BrandController {
	
	@Autowired
	private BrandService bservice;
	
	@PostMapping(value="/add",consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
	public ResponseEntity<?> addBrand(@RequestParam String brandName,
			       @RequestParam String about , @RequestParam(value = "status",defaultValue="ACTIVE") String status,
			       @RequestPart(value="image",required = false) MultipartFile image){
//		
//		if(result.hasErrors()) {
//			throw new AppException(result.getFieldError().getDefaultMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
//		}
		
		AddBrandRequest request = new AddBrandRequest();
		request.setBrandName(brandName);
		request.setAbout(about);
		request.setStatus(status);
		BrandDto dto = bservice.addBrand(request, image);
		return ResponseEntity.ok(new ApiResponse<>("brand added sucessfully!",dto,HttpStatus.OK));
	}

	@GetMapping("/get/{brandId}")
	public ResponseEntity<?> getBrandById(@PathVariable Integer brandId){
		BrandDto dto = bservice.getById(brandId);
		return ResponseEntity.ok(new ApiResponse<>("brand data!",dto,HttpStatus.OK));
	}
	
	@GetMapping("/getAll")
	public ResponseEntity<?> getAllBrands(){
		List<BrandDto> list = bservice.getAllBrands();
		return ResponseEntity.ok(new ApiResponse<>("brand data!",list,HttpStatus.OK));
	}
	
	@PutMapping(value="/update/{brandId}",consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
	public ResponseEntity<?> updateBrand(@PathVariable Integer brandId,@RequestParam String brandName,@RequestParam String about , @RequestParam String status,@RequestPart(value="image",required = false) MultipartFile image){
		
//		if(result.hasErrors()) {
//			throw new AppException(result.getFieldError().getDefaultMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
//		}
		
		UpdateBrandRequest request = new UpdateBrandRequest();
		request.setBrandName(brandName);
		request.setAbout(about);
		request.setStatus(status);
		BrandDto dto = bservice.updateBrand(brandId, request, image);
		return ResponseEntity.ok(new ApiResponse<>("updated sucessfully!",dto,HttpStatus.OK));
	}
	
	@DeleteMapping("/delete/{brandId}")
	public ResponseEntity<?> deleteBrand(@PathVariable Integer brandId){
		bservice.deleteBrandById(brandId);
		return ResponseEntity.ok(new ApiResponse<>("deleted sucessfully!",null,HttpStatus.OK));
	}
	
	
	
	
	
	
	
	
}
